Hrm, these tests are passing locally for me:
```
E         -  'array': {'data': 'H4sIAAAAAAACE2NgYGBgBGImIGYGYhYgZgViAD34DIUYAAAA',
E         ?                                 ^
E         +  'array': {'data': 'H4sIAAAAAAAC/2NgYGBgBGImIGYGYhYgZgViAD34DIUYAAAA',
E         ?     
```
Perusing the [Gzip RFC](https://www.rfc-editor.org/rfc/rfc1952.html#section-2.2) it seems as though an "OS" field is encoded in the header, which I will speculate is causing the difference here. So it looks like those tests will have to be handled a little more carefully... Since these are on the python side it might just be easier to round-trip everything for comparison, rather than comparing the encoded values directly. 

Edit: really strange/interesting, OSX tests for 3.11 and 3.12 pass, which is consistent with my running locally on OSX. But OSX for 3.13 is *not* passing. Definitely seems like all these will just have to round-trip everything.

--- a/src/bokeh/core/serialization.py
+++ b/src/bokeh/core/serialization.py
@@ -45,6 +45,7 @@ from typing import (
 import numpy as np
 
 # Bokeh imports
+from ..settings import settings
 from ..util.dataclasses import (
     Unspecified,
     dataclass,
@@ -176,11 +177,12 @@ class Buffer:
         return self.data.tobytes() if isinstance(self.data, memoryview) else self.data
 
     def to_compressed_bytes(self) -> bytes:
+        level = settings.compression_level()
         # we do not want the result to be different depending on mtime, since that is
         # irrelevant and also makes things harder to test, but Python 3.11 and 3.12 have
         # bug where using mtime=0 results in the Gzip header OS field varies by platforam
         # instead of getting set to a fixed value of 255. So, for now use mtime=1 instead.
-        return gzip.compress(self.to_bytes(), mtime=1)
+        return gzip.compress(self.to_bytes(), mtime=1, compresslevel=level)
 
     def to_base64(self) -> str:
         return base64.b64encode(self.to_compressed_bytes()).decode("utf-8")

--- a/src/bokeh/core/serialization.py
+++ b/src/bokeh/core/serialization.py
@@ -177,8 +177,10 @@ class Buffer:
 
     def to_compressed_bytes(self) -> bytes:
         # we do not want the result to be different depending on mtime, since that is
-        # irrelevant and also makes things harder to test, so set mtime=0 here
-        return gzip.compress(self.to_bytes(), mtime=0)
+        # irrelevant and also makes things harder to test, but Python 3.11 and 3.12 have
+        # bug where using mtime=0 results in the Gzip header OS field varies by platforam
+        # instead of getting set to a fixed value of 255. So, for now use mtime=1 instead.
+        return gzip.compress(self.to_bytes(), mtime=1)
 
     def to_base64(self) -> str:
         return base64.b64encode(self.to_compressed_bytes()).decode("utf-8")

--- a/src/bokeh/settings.py
+++ b/src/bokeh/settings.py
@@ -169,6 +169,16 @@ def convert_int(value: int | str) -> int:
     '''
     return int(value)
 
+def convert_compression(value: int | str) -> int:
+    ''' Convert a string to a gzip compression level value.
+    '''
+    level = int(value)
+
+    if 0 <= level <= 9:
+        return level
+
+    raise ValueError(f"Compression level must be an integer in [0, 9], got {value!r}")
+
 def convert_bool(value: bool | str) -> bool:
     ''' Convert a string to True or False.
 
@@ -499,6 +509,8 @@ class PrioritizedSetting(Generic[T]):
             return "Bool"
         if self._convert is convert_int:
             return "Int"
+        if self._convert is convert_compression:
+            return "Compression Level (0-9)"
         if self._convert is convert_logging:
             return "Log Level"
         if self._convert is convert_str_seq:
@@ -581,6 +593,16 @@ class Settings:
     different name for ``chromedriver``, like ``chromedriver-binary`` or
     ``chromium.chromedriver`` (or its variant, which is used for example
     by Snap package manager; see https://snapcraft.io/).
+    """)
+
+    compression_level: PrioritizedSetting[int] = PrioritizedSetting("compression_level", "BOKEH_COMPRESSION_LEVEL", default=9, convert=convert_compression, help="""
+    In contexts where array buffers are base64-encoded (e.g. to embed inside
+    an HTML file), the buffer will first be compressed to save space.
+
+    Valid values are the standard gzip compression levels 0-9. A setting of 9
+    (the default) will result in the highest compression. A setting of 1 will
+    result in the least compression, but be faster. A setting of 0 will result
+    in no compression.
     """)
 
     cookie_secret: PrioritizedSetting[str | None] = PrioritizedSetting("cookie_secret", "BOKEH_COOKIE_SECRET", default=None, help="""

--- a/tests/unit/bokeh/core/test_json_encoder.py
+++ b/tests/unit/bokeh/core/test_json_encoder.py
@@ -24,6 +24,7 @@ from math import nan
 # Bokeh imports
 from bokeh.core.json_encoder import serialize_json
 from bokeh.core.serialization import Serializer
+from bokeh.settings import settings
 
 #-----------------------------------------------------------------------------
 # Setup
@@ -87,7 +88,8 @@ def test_json_encoder_bytes():
     assert rep.buffers is not None and len(rep.buffers) == 1
 
     # Note mtime=1 is a workaround for an issue with Python 3.11 and 3.12
-    encoded_bytes = b64encode(gzip.compress(val["key"], mtime=1)).decode("utf-8")
+    compressed_bytes = gzip.compress(val["key"], mtime=1, compresslevel=settings.compression_level())
+    encoded_bytes = b64encode(compressed_bytes).decode("utf-8")
 
     assert serialize_json(rep.content) == f"""\
 {{"type":"map","entries":[["key",{{"type":"bytes","data":"{encoded_bytes}"}}]]}}\

--- a/tests/unit/bokeh/core/test_json_encoder.py
+++ b/tests/unit/bokeh/core/test_json_encoder.py
@@ -17,6 +17,8 @@ import pytest ; pytest
 #-----------------------------------------------------------------------------
 
 # Standard library imports
+import gzip
+from base64 import b64encode
 from math import nan
 
 # Bokeh imports
@@ -32,17 +34,17 @@ from bokeh.core.serialization import Serializer
 #-----------------------------------------------------------------------------
 
 def test_json_encoder():
-    val0 = [None, True, False, -128, -1, 0, 1, 128, nan, {"key_0": b"uvw"}]
+    val0 = [None, True, False, -128, -1, 0, 1, 128, nan]
     rep0 = Serializer().serialize(val0)
 
-    assert rep0.buffers is not None and len(rep0.buffers) == 1
+    assert rep0.buffers is not None and len(rep0.buffers) == 0
 
     assert serialize_json(rep0.content) == """\
-[null,true,false,-128,-1,0,1,128,{"type":"number","value":"nan"},{"type":"map","entries":[["key_0",{"type":"bytes","data":"H4sIAAAAAAACEystKwcARkl/GgMAAAA="}]]}]\
+[null,true,false,-128,-1,0,1,128,{"type":"number","value":"nan"}]\
 """
 
-    assert serialize_json(rep0) == f"""\
-[null,true,false,-128,-1,0,1,128,{{"type":"number","value":"nan"}},{{"type":"map","entries":[["key_0",{{"type":"bytes","data":{{"id":"{rep0.buffers[0].id}"}}}}]]}}]\
+    assert serialize_json(rep0) == """\
+[null,true,false,-128,-1,0,1,128,{"type":"number","value":"nan"}]\
 """
 
     assert serialize_json(rep0.content, pretty=True) == """\
@@ -58,23 +60,11 @@ def test_json_encoder():
   {
     "type": "number",
     "value": "nan"
-  },
-  {
-    "type": "map",
-    "entries": [
-      [
-        "key_0",
-        {
-          "type": "bytes",
-          "data": "H4sIAAAAAAACEystKwcARkl/GgMAAAA="
-        }
-      ]
-    ]
   }
 ]\
 """
 
-    assert serialize_json(rep0, pretty=True) == f"""\
+    assert serialize_json(rep0, pretty=True) == """\
 [
   null,
   true,
@@ -84,25 +74,22 @@ def test_json_encoder():
   0,
   1,
   128,
-  {{
+  {
     "type": "number",
     "value": "nan"
-  }},
-  {{
-    "type": "map",
-    "entries": [
-      [
-        "key_0",
-        {{
-          "type": "bytes",
-          "data": {{
-            "id": "{rep0.buffers[0].id}"
-          }}
-        }}
-      ]
-    ]
-  }}
+  }
 ]\
+"""
+
+def test_json_encoder_bytes():
+    val = {"key": b"uvw"}
+    rep = Serializer().serialize(val)
+    assert rep.buffers is not None and len(rep.buffers) == 1
+
+    encoded_bytes = b64encode(gzip.compress(val["key"], mtime=0)).decode("utf-8")
+
+    assert serialize_json(rep.content) == f"""\
+{{"type":"map","entries":[["key",{{"type":"bytes","data":"{encoded_bytes}"}}]]}}\
 """
 
 def test_json_encoder_dict_no_sort():

--- a/tests/unit/bokeh/core/test_json_encoder.py
+++ b/tests/unit/bokeh/core/test_json_encoder.py
@@ -86,7 +86,8 @@ def test_json_encoder_bytes():
     rep = Serializer().serialize(val)
     assert rep.buffers is not None and len(rep.buffers) == 1
 
-    encoded_bytes = b64encode(gzip.compress(val["key"], mtime=0)).decode("utf-8")
+    # Note mtime=1 is a workaround for an issue with Python 3.11 and 3.12
+    encoded_bytes = b64encode(gzip.compress(val["key"], mtime=1)).decode("utf-8")
 
     assert serialize_json(rep.content) == f"""\
 {{"type":"map","entries":[["key",{{"type":"bytes","data":"{encoded_bytes}"}}]]}}\

--- a/tests/unit/bokeh/core/test_serialization.py
+++ b/tests/unit/bokeh/core/test_serialization.py
@@ -25,6 +25,7 @@ from base64 import b64decode
 from math import inf, nan
 from types import SimpleNamespace
 from typing import Any, Sequence
+from unittest.mock import patch
 
 # External imports
 import numpy as np
@@ -98,6 +99,74 @@ class SomeDataClass:
     f3: NotRequired[bool | None] = Unspecified
     f4: NotRequired[SomeProps] = Unspecified
     f5: NotRequired[SomeModel] = Unspecified
+
+class TestBuffer:
+
+    def test_ref(self) -> None:
+        buf = Buffer(10, b"\x001\xefabc\x12")
+        assert buf.ref == {'id': 10}
+
+    def test_to_bytes_from_bytes(self) -> None:
+        val = b"\x001\xefabc\x12"
+        buf = Buffer(10, val)
+        ret = buf.to_bytes()
+        assert isinstance(ret, bytes)
+        assert ret == val
+
+    def test_to_bytes_from_memoryview(self) -> None:
+        val = b"\x001\xefabc\x12"
+        view = memoryview(val)
+        buf = Buffer(10, view)
+        ret = buf.to_bytes()
+        assert isinstance(ret, bytes)
+        assert ret == val
+
+    def test_to_compressed_bytes_from_bytes(self) -> None:
+        val = b"\x001\xefabc\x12"
+        buf = Buffer(10, val)
+
+        ret = buf.to_compressed_bytes()
+
+        assert isinstance(ret, bytes)
+
+        # make sure the gzip header "OS" field is *always* set to "unknown" (255)
+        assert ret[9] == 255
+
+        assert gzip.decompress(ret) == val
+
+    def test_to_compressed_bytes_from_memoryview(self) -> None:
+        val = b"\x001\xefabc\x12"
+        view = memoryview(val)
+        buf = Buffer(10, view)
+
+        ret = buf.to_compressed_bytes()
+
+        assert isinstance(ret, bytes)
+
+        # make sure the gzip header "OS" field is *always* set to "unknown" (255)
+        assert ret[9] == 255
+
+        assert gzip.decompress(ret) == val
+
+    @pytest.mark.parametrize("level", range(0, 10))
+    @patch("gzip.compress")
+    def test_to_compressed_compression_level(self, compress, level: int, monkeypatch: pytest.MonkeyPatch) -> None:
+        val = b"\x001\xefabc\x12"
+        buf = Buffer(10, val)
+
+        monkeypatch.setenv("BOKEH_COMPRESSION_LEVEL", f"{level}")
+        buf.to_compressed_bytes()
+
+        compress.assert_called_once_with(val, mtime=1, compresslevel=level)
+
+    def test_to_base64(self):
+        val = b"\x001\xefabc\x12"
+        buf = Buffer(10, val)
+
+        ret = buf.to_base64()
+
+        assert isinstance(ret, str)
+        assert b64decode(ret) == buf.to_compressed_bytes()
 
 class TestSerializer:
 

--- a/tests/unit/bokeh/core/test_serialization.py
+++ b/tests/unit/bokeh/core/test_serialization.py
@@ -18,8 +18,10 @@ import pytest ; pytest
 
 # Standard library imports
 import datetime as dt
+import gzip
 import sys
 from array import array as TypedArray
+from base64 import b64decode
 from math import inf, nan
 from types import SimpleNamespace
 from typing import Any, Sequence
@@ -309,10 +311,8 @@ class TestSerializer:
         encoder = Serializer(deferred=False)
         val = bytes([0xFF, 0x00, 0x17, 0xFE, 0x00])
         rep = encoder.encode(val)
-        assert rep == BytesRep(
-            type="bytes",
-            data="H4sIAAAAAAACE/vPIP6PAQBXSRCfBQAAAA==",
-        )
+        assert rep["type"] == "bytes"
+        assert gzip.decompress(b64decode(rep["data"])) == val
         assert encoder.buffers == []
 
     def test_typed_array(self) -> None:
@@ -336,15 +336,13 @@ class TestSerializer:
         encoder = Serializer(deferred=False)
         val = TypedArray("i", [0, 1, 2, 3, 4, 5])
         rep = encoder.encode(val)
-        assert rep == TypedArrayRep(
-            type="typed_array",
-            array=BytesRep(
-                type="bytes",
-                data="H4sIAAAAAAACE2NgYGBgBGImIGYGYhYgZgViAD34DIUYAAAA",
-            ),
-            order=sys.byteorder,
-            dtype="int32",
-        )
+        # isinstance not supported for typed dicts, alas
+        # assert isinstance(rep, TypedArrayRep)
+        assert rep.keys() == {"type", "order", "dtype", "array"}
+        assert rep["type"] == "typed_array"
+        assert rep["order"] == sys.byteorder
+        assert rep["dtype"] == "int32"
+        assert rep["array"] == encoder._encode_bytes(memoryview(val))
         assert encoder.buffers == []
 
     def test_ndarray(self) -> None:
@@ -369,16 +367,14 @@ class TestSerializer:
         encoder = Serializer(deferred=False)
         val = np.array([0, 1, 2, 3, 4, 5], dtype="int32")
         rep = encoder.encode(val)
-        assert rep == NDArrayRep(
-            type="ndarray",
-            array=BytesRep(
-                type="bytes",
-                data="H4sIAAAAAAACE2NgYGBgBGImIGYGYhYgZgViAD34DIUYAAAA",
-            ),
-            order=sys.byteorder,
-            shape=[6],
-            dtype="int32",
-        )
+        # isinstance not supported for typed dicts, alas
+        # assert isinstance(rep, NDArrayRep)
+        assert rep.keys() == {"type", "order", "shape", "dtype", "array"}
+        assert rep["type"] == "ndarray"
+        assert rep["order"] == sys.byteorder
+        assert rep["shape"] == [6]
+        assert rep["dtype"] == "int32"
+        assert rep["array"] == encoder._encode_bytes(memoryview(val))
         assert encoder.buffers == []
 
     def test_ndarray_dtypes_shape(self) -> None:

--- a/tests/unit/bokeh/test_settings.py
+++ b/tests/unit/bokeh/test_settings.py
@@ -48,6 +48,7 @@ _expected_settings = (
     'browser',
     'cdn_version',
     'chromedriver_path',
+    'compression_level',
     'cookie_secret',
     'default_server_host',
     'default_server_port',
@@ -99,6 +100,8 @@ class TestSettings:
 
         assert bs.settings.default_server_port.convert_type == "Int"
 
+        assert bs.settings.compression_level.convert_type == "Compression Level (0-9)"
+
         assert bs.settings.py_log_level.convert_type == "Log Level"
 
         assert bs.settings.validation_level.convert_type == "Validation Level"
@@ -108,15 +111,16 @@ class TestSettings:
         assert bs.settings.ico_path.convert_type == "Ico Path"
 
         default_typed = set(_expected_settings) - {
+            'allowed_ws_origin',
+            'compression_level',
+            'default_server_port',
             'ico_path',
             'ignore_filename',
             'minified',
-            'default_server_port',
             'perform_document_validation',
+            'py_log_level',
             'simple_ids',
             'validation_level',
-            'py_log_level',
-            'allowed_ws_origin',
             'xsrf_cookies',
         }
         for name in default_typed:

